docs(x-integration): review build search UI docs (#423)
empathyco/x: packages/x-components/static-docs/build-search-ui/web-archetype-integration-guide.md

tags:
  - integration
  - archetype
  - archetype integration
  - x integration
  - interface x
  - x components
---

# Integrate Interface X Archetype into an existing website

In this tutorial, you'll learn how to integrate the Interface&nbsp;X&nbsp;Archetype project in your commerce store in a matter of minutes. You can use the X&nbsp;Archetype as is or you can [extend the search and discovery interface experience](web-archetype-development-guide.md) to meet your business needs.

To integrate the Interface&nbsp;X&nbsp;Archetype layer in your commerce store, just **load** the generated Interface&nbsp;X JavaScript file and **initialize** it.

> 

Reviewer: We can include here a info Panel with info like this:

If you are loading an archetype script hosted by empathy, X resources are served by a CDN through the following URLs Environment:

- Live: https://x.empathy.co/{INSTANCE}/app.js
- Staging: https://x.staging.empathy.co/{INSTANCE}/app.js

where {INSTANCE} is the name of your store accorded with Empathy.

Author: Just curious, has this instance something to do with the instances a partner can configure with the CMC tool from Playboard, @herrardo? See https://docs-dev.empathy.co/explore-empathy-platform/configure-empathy-platform/request-instance.html

Reviewer: I think so, you create the instance, or ask to Empathy to create it for you in fact, and then you set here the InstanceId, but i wouldn't distract this developer with things about the playboard, because Playboard is not a thing that this Frontend integrator developer is going to manage IMO

Author: Included as a note including a call to support in case of help requesting the instance.

  };
};
```
::: note

> You can cange the snippet configuration values once the project is deployed. Use the `/x-archetype/public/snippet-script.js` file to perform hot changes for the snippet parameters. For more information on the supported parameters, check out [Snippet configuration](#snippet-configuration).

Reviewer: ```suggestion
You can change the snippet configuration values once the project is deployed. Use the `/x-archetype/public/snippet-script.js` file to perform hot changes for the snippet parameters. For more information on the supported parameters, check out [Snippet configuration](#snippet-configuration).
```

</script>
```

For this example, the initialization function is called immediately after loading the script, but it can be called at any time. Note that you need to call this function only **once**.

> <a name="callbacks"></a>

Reviewer 2: What is this empty link?

::: interact

> Check out the [X&nbsp;API](#x-api) section to learn more about the functions and parameters supported.

Reviewer: You have removed the Required column in the new table, and it is much less clear.

Author: I've removed the _Required_ column to both reduce the length of the overall table and enlarge the width of the _Description_ column. Having too long descriptions makes it difficult to read. However, the information on which param is required or not is still available in the table. This information is not missing but met the same style used in the UI reference tables.  
Due to the huge size of descriptions, I've also relocated some additional information, by removing it from the table and placing it in callouts or other related sections. I've not removed nor deleted any important information, just rewritten or relocated it.

:::

> ::: develop Initialization using callbacks

Reviewer: You removed the Callbacks section, so now is part of `2. Initialize Interface X` inside `Initializing Interface X project on demand`, but this is not correct, because you can set the callbacks also in `Initializing Interface X project automatically``

I liked more the way it was before, but if you want to separate it, we should have a menu like:

```
## Further considerations

### Snippet Configuration
### Setting callbacks up
### Events
### X API public functions
````

Also the paragraph at the beggining of the current Further Considerations with the three bullets is unnecesary because you are about to read each section. I feel there ir more links and info, but less clear now in this part.

Reviewer: FURTHER CONSIDERATIONS neither seems the best title here to me.

| `searchLang`              | `string`     | _Optional_. Language to use for the API requests **only**.  |
| `consent`                | `boolean`    |_Required._ Determines whether the shopper has accepted the use of cookies so that the `sessionId` is sent to the Empathy's Search and Tagging APIs or not.  |
| `documentDirection`       | `'ltr'` &#124; `'rtl'`    |_Optional_. Writing direction script that the X Components should, i.e. left-to-right or right-to-left.   |
| `currency`                | `string`   |_Required._ Currency identifier to configure how prices are displayed.  |
| [`callbacks`](#callbacks-and-interface-x-events-types) | `Record<XEventName, (payload: XEventPayload<Event>, metadata: WireMetadata) => void` | _Optional_.   Callback record where the _key_ is the event to listen and the _value_ is the callback to be executed whenever the event is emitted. E.g. to listen to the `UserAcceptedAQuery` event: `{ UserAcceptedAQuery({ eventPayload }) { console.log('UserAcceptedAQuery', eventPayload); }`  |

> | `isSpa`                   | `boolean`    |_Optional_. Enables signgle-page application model. You set it to `true` when the X&nbsp;Archetype runs on top of a SPA website.   |

Reviewer 2: ```suggestion
| `isSpa`                   | `boolean`    |_Optional_. Enables single-page application model. You set it to `true` when the X&nbsp;Archetype runs on top of a SPA website.   |
```

</script>
```

For this example, the initialization function is called immediately after loading the script, but it can be called at any time. Note that you need to call this function only **once**.

<a name="callbacks"></a>

::: interact

Check out the [X&nbsp;API](#x-api) section to learn more about the functions and parameters supported.

:::

## Notes on X Archetype integration

To successfully integrate Interface&nbsp;X in your commerce store using the X&nbsp;Archetype, check out further information about:

* **Initialization options** supported in [snippet configuration](#snippet-configuration)

> * **[Callbacks and X&nbsp;event&nbsp;types](#callbacks-interface-x-events-types)** available to subscribe to when initializing

Author: Guess we're having a linking conflict here among repos. For the link to work properly in code needs the ampersand, but it won't work in eDocs. I'll replace the `&` symbol to avoid conflicts.

> * **[Callbacks and X&nbsp;event&nbsp;types](#callbacks-interface-x-events-types)** available to subscribe to when initializing

Author: Now works in my local and in eDocs

| `env`                     | `'live'` &#124; `'staging'`              | API environment to use. You can use the Interface&nbsp;X  production version with the staging API, and vice versa.  |
| `scope`                   | `string`    | Context where the search interface is executed, i.e. `mobile`, `mobile-app`, `tablet`, `desktop`.  |
| `lang`                    | `string`                                                                             | _Required._ Language to use. By default, it's used for both the frontend and the API requests.   |
| `searchLang`              | `string`     | Language to use for the API requests **only**.  |
| `consent`                | `boolean`    |_Required._ Determines whether the shopper has accepted the use of cookies so that the `sessionId` is sent to the Empathy's Search and Tagging APIs or not.  |
| `documentDirection`       | `'ltr'` &#124; `'rtl'`    | Writing direction script that the X Components should, i.e. left-to-right or right-to-left.   |
| `currency`                | `string`   |_Required._ Currency identifier to configure how prices are displayed.  |

> | [`callbacks`](#callbacks-interface-x-events-types) | `Record<XEventName, (payload: XEventPayload<Event>, metadata: WireMetadata) => void` |    Callback record where the _key_ is the event to listen and the _value_ is the callback to be executed whenever the event is emitted. E.g. to listen to the `UserAcceptedAQuery` event: `{ UserAcceptedAQuery({ eventPayload }) { console.log('UserAcceptedAQuery', eventPayload); }  |

Author: Fixed. It works in my local and in eDocs

> | `isSpa`                   | `boolean`    | Defines signgle-page application model. You set to `true` when the X&nbsp;Archetype runs on top of an SPA website.   |

Reviewer 2: ```suggestion
| `isSpa`                   | `boolean`    | Enables single-page application model. You set to `true` when the X&nbsp;Archetype runs on top of an SPA website.   |
```

    }
  });
</script>
```
``` 
:::

In this example we are subscribing to `UserAcceptedAQuery` and `UserClickedResultAddToCart` so in
this way we can perform an action inside our application when these events are triggered. Even
though we have subscribed to two events, as Interface X Components is an event based architecture
there are more than one hundred of events available to be subscribed to. The complete list is not
yet part of this documentation, but you may access to the
[XEventsTypes](https://github.com/empathyco/x/blob/main/packages/x-components/src/wiring/events.types.ts)

> where all of them are defined.

Reviewer: Why are we removing this explanation? I know there are things that you are telling ion other side, but not all of them.

Author: I think I've removed nothing, but rewritten or replaced information. Now, callout information is now available as a stand-alone subsection covering the example and the info related to X events as well.

| `env`                     | `'live'` &#124; `'staging'`              | API environment to use. You can use the Interface&nbsp;X  production version with the staging API, and vice versa.  |
| `scope`                   | `string`    | Context where the search interface is executed, i.e. `mobile`, `mobile-app`, `tablet`, `desktop`.  |
| `lang`                    | `string`                                                                             | _Required._ Language to use. By default, it's used for both the frontend and the API requests.   |
| `searchLang`              | `string`     | Language to use for the API requests **only**.  |
| [`consent`](#consent)                | `boolean`    |_Required._ Determines whether the shopper has accepted the use of cookies so that the `sessionId` is sent to the Empathy's Search and Tagging APIs or not.  |
| `documentDirection`       | `'ltr'` &#124; `'rtl'`    | Writing direction script that the X Components should, i.e. left-to-right or right-to-left.   |
| `currency`                | `string`   |_Required._ Currency identifier to configure how prices are displayed.  |

> | [`callbacks`](#callbacks) | `Record<XEventName, (payload: XEventPayload<Event>, metadata: WireMetadata) => void` |    Callback record where the _key_ is the event to listen and the _value_ is the callback to be executed whenever the event is emitted. E.g. to listen to the `UserAcceptedAQuery` event: `{ UserAcceptedAQuery({ eventPayload }) { console.log('UserAcceptedAQuery', eventPayload); }  |

Reviewer 2: Are callback and consent links working? I can't find their target, but they seem to link to a random place in the document.

Reviewer: There was a callbacks section but Luisa got rid of it, but we are going to restore it, righ @mlcorton ? :)

Author: My fault! Sorry, I misunderstood the info provided! I'll revert the initial structure

#### X API

The [X&nbsp;API](https://github.com/empathyco/x/blob/main/packages/x-components/src/x-installer/api/base-api.ts) object allows your commerce store to communicate with Interface&nbsp;X. It supports multiple functions to integrate Interface&nbsp;X in your website.

| Function         | Parameters                                                                     | Description                                                                                                                                                       |
| ---------------- | ------------------------------------------------------------------------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `init`             | [snippet configuration params](#snippet-configuration) - Initialization options | [Initializes Interface&nbsp;X on demand](#initializing-interface-x-project-on-demand).   |
| `search`           | `query` - Query to open Interface&nbsp;X  | _Optional_. Executes Interface&nbsp;X and triggers a search with the given query.  |

> | `setSnippetConfig` | [snippet configuration params](#snippet-configuration) - Initialization options | Changes initialization options so that all components react to the changes, i.e. changing both search engine and language without reloading the page. |

Reviewer: ```suggestion
| `setSnippetConfig` | [snippet configuration params](#snippet-configuration) - Initialization options | Changes initialization options so that all components react to the changes, i.e. changing both search engine and language without reloading the page. |
| `addProductToCart` | none | To be called from PDP (Product Detail Page) when the user clicks on the Add To Cart Button. Calling this function will send tracking of the AddToCart event to the [Empathy Tagging API](https://docs.empathy.co/develop-empathy-platform/capture-interaction-signals/tagging-api-guide.html) for the product which is currently in screen.  |
```

I am adding this other event because yesterday we realized we didn't document it, and this is basic to integrate customers. cc @javieri-empathy

Author: To follow the same structure and style before, I'll rewrite the description field: 

> Sends tracking of the `AddToCart`event to the Empathy Tagging microservice for the selected /displayed product.

Can't commit your suggestion directly inhere. If you agree with the proposed description, I'll add the new info in a new commit together with the rest of changes.

Author: To follow the same structure and style before, I'll rewrite the description field: 

> Sends tracking of the `AddToCart`event to the Empathy Tagging microservice for the selected /displayed product.

Can't commit your suggestion directly inhere. If you agree with the proposed description, I'll added the new info in a new commit together with the rest of changes.

Reviewer 2: I would just add that this function might have a parameter: the product id. If it is not provided X Components will try to use the url as a identifier, but it might fail.

| Function         | Parameters                                                                     | Description                                                                                                                                                       |
| ---------------- | ------------------------------------------------------------------------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `init`             | [snippet configuration params](#snippet-configuration) - Initialization options | [Initializes Interface&nbsp;X on demand](#initializing-interface-x-project-on-demand).   |
| `search`           | `query` - Query to open Interface&nbsp;X  | _Optional_. Executes Interface&nbsp;X and triggers a search with the given query.  |
| `setSnippetConfig` | [snippet configuration params](#snippet-configuration) - Initialization options | Changes initialization options so that all components react to the changes, i.e. changing both search engine and language without reloading the page. |

> | `addProductToCart` | none | Sends tracking of the `AddToCartevent` to the [Empathy Tagging microservice](https://docs.empathy.co/develop-empathy-platform/capture-interaction-signals/tagging-api-guide.html) for the product displayed on screen. This function is called from  the product detail page (PDP) when the shopper clicks on the add-to-cart button. |

Reviewer 2: ```suggestion
| `addProductToCart` | `productId` __Optional__ - The product identifier that has been added to the cart  | Sends tracking of the `AddToCart` event to the [Empathy Tagging microservice](https://docs.empathy.co/develop-empathy-platform/capture-interaction-signals/tagging-api-guide.html) for the product displayed on screen. This function is called from  the product detail page (PDP) when the shopper clicks on the add-to-cart button. If the `productId` is not provided, the URL will be used to detect if the user has discovered the product thanks to a search session. |
```

Author: Just wondering @herrardo @javieri-empathy, if the parameters for `AddProductToCart` and `Search` functions are optional, then do I guess that parameters for other functions (`setSnippetConfig` or `Init`) are required?? If so, should we say which are required rather than the optional ones? I'm just trying to unify styles, in the Config snippet table and the Props/Slots in UI ref, we indicate the required ones only.

Reviewer: My answer is going to hurt you. I would inform every one of them with _Optional_ or _Required_because when it was a check on a column, it was obvious, but now it is not so.

Anyway, if you inform only about the Required ones is not a stopper for me.
